joomla/joomla-cms
# Add version and IE conditional statements to JHtml::stylesheet/JDocument::addStyleSheet methods and deprecate JDocument::addSyleSheetVersion()

Pull Request for Improvement.
### Summary of Changes

This PR does for stylesheets what was done for script in https://github.com/joomla/joomla-cms/pull/11289

Ie, adds the ability to add add media version and IE conditional statements to the stylesheets link tag using `JDocument::addStyleSheet()` and `JHtml::stylesheet()` methods.  

Also deprecates `JDocument::addSyleSheetVersion()` method.
### Testing Instructions

Please test this in all possible scenarios.
Any problem will get stylesheets not loading.
- Do a code review.
- Use 3.7.x branch and apply patch
- Check backend and frontend styles are working fine
- Add external css stylesheets like to isis/protostar index.php, some examples:

``` php
// JDocument B/C tests
/* addStyleSheet($url, $type = 'text/css', $media = null, $attribs = array()) */
$this->addStyleSheet('/path/to/external/bcaddStyleSheet1.css');
$this->addStyleSheet('/path/to/external/bcaddStyleSheet2.css', '', null, array());
$this->addStyleSheet('/path/to/external/bcaddStyleSheet3.css', 'text/css', 'all', array('scriptid' => 1));
/* addStyleSheetVersion($url, $version = null, $type = 'text/css', $media = null, $attribs = array()) */
$this->addStyleSheetVersion('/path/to/external/bcaddStyleSheetVersion1.css');
$this->addStyleSheetVersion('/path/to/external/bcaddStyleSheetVersion2.css', null, '', null, array());
$this->addStyleSheetVersion('/path/to/external/bcaddStyleSheetVersion3.css', 'version-string', 'text/css', 'all', array('scriptid' => 1));

// JDocument new method signature tests
/* addStyleSheet($url, $options = array(), $attribs = array()) */
$this->addStyleSheet('/path/to/external/addStyleSheet0.css');
$this->addStyleSheet('/path/to/external/addStyleSheet1.css', array(), array('id' => 'stylesheetid'));
$this->addStyleSheet('/path/to/external/addStyleSheet2.css', array(), array('id' => 'stylesheetid', 'data-test' => 1));
$this->addStyleSheet('/path/to/external/addStyleSheet3.css', array(), array('data-attrib' => 'original-value'));
$this->addStyleSheet('/path/to/external/addStyleSheet4.css', array(), array('data-attrib' => 'final-value'));
$this->addStyleSheet('/path/to/external/addStyleSheet5.css', array(), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addStyleSheet('/path/to/external/addStyleSheet6.css', array('version' => 'auto'), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addStyleSheet('/path/to/external/addStyleSheet7.css', array('version' => 'version-string'), array('media' => 'all', 'data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));
$this->addStyleSheet('/path/to/external/addStyleSheet8.css', array('version' => 'auto', 'conditional' => 'lf IE 9'), array('data-attrib' => array('item1' => 'value1', 'item2' => 'value2')));

// JHTML B/C Tests
/* stylesheet($file, $attribs = array(), $relative = false, $path_only = false, $detect_browser = true, $detect_debug = true) */
JHtml::_('stylesheet', 'media/system/css/jquery.Jcrop.min.css');
JHtml::_('stylesheet', 'system/frontediting.css', array('id' => 'stylesheetid'), true, false, false, false);

// JHtml new method signature tests
/* stylesheet($file, $options = array(), $attribs = array()) */
JHtml::_('stylesheet', 'media/com_finder/css/indexer.css', array(), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('stylesheet', 'system/adminlist.css', array('relative' => true), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('stylesheet', 'system/calendar-jos.css', array('relative' => true, 'version' => 'auto'), array('data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('stylesheet', 'system/mootree.css', array('relative' => true, 'version' => 'version-string'), array('media' => 'all', 'data-attrib' => array('item1' => 3, 'item2' => 'value2')));
JHtml::_('stylesheet', 'system/mootree_rtl.css', array('relative' => true, 'version' => 'version-string', 'conditional' => 'lf IE 9'), array('media' => 'all', 'data-attrib' => array('item1' => 3, 'item2' => 'value2')));
```
- Check the HTML output and confirm all is fine
- Do the same tests for anything else you can remember
### Observations

From a B/C perspective i think there are no issues, but mantainers please check.
